Downstream only: Rework of Makefile and incusion of lint

mpryc opened the pull request:
The rework of Makefile to make it more readable and inclusion of lint as a target.

- Removed unnecessary installation of `code-generator` which was not used in the Makefile.prow
- Removed copy of the code to separate folder, which was required by `code-generator` 
- Removed installation of kubectl, which isn't used at the moment by unit and lint jobs
- Rework of code to be eye-candy
- KUBEBUILDER_ASSETS moved inside test target, so it's not invoked for lint job
- Added fake `verify-modules` target so now `make ci` is calling exactly same targets as upstream `Makefile`
- Replaced docker with podman for running tests locally as docker is not available by default in RHEL/Fedora.

## Tests performed:
```shell
$ podman run -it --rm -v `pwd`:`pwd`:Z -w `pwd` quay.io/konveyor/builder:ubi9-v1.22.2 bash -c "make -f Makefile.prow ci"
$ podman run -it --rm -v `pwd`:`pwd`:Z -w `pwd` quay.io/konveyor/builder:ubi9-v1.22.2 bash -c "make -f Makefile.prow lint"
```

mateusoliveira43 commented on Makefile.prow:
```diff
@@ -1,92 +1,87 @@
-# 
-# If formatting needs updating you can run
+# Warning: Verify this inside a container with the following command:
+# podman run -it --rm -v `pwd`:`pwd`:Z -w `pwd` golang:1.22 bash -c "make -f Makefile.prow ci"
```
I think makes more sense to use same image used by CI https://github.com/openshift/release/blob/master/ci-operator/config/openshift/velero/openshift-velero-konveyor-dev.yaml#L5

mpryc replied:
You are right I will add the following two (I used as well :Z to ensure code is readable with SELinux):
```
# podman run -it --rm -v `pwd`:`pwd`:Z -w `pwd` quay.io/konveyor/builder:ubi9-v1.22.2 bash -c "make -f Makefile.prow ci"
# podman run -it --rm -v `pwd`:`pwd`:Z -w `pwd` quay.io/konveyor/builder:ubi9-v1.22.2 bash -c "make -f Makefile.prow lint"
```

mateusoliveira43 commented on Makefile.prow:
```diff
@@ -1,92 +1,87 @@
+
+.PHONY: verify-modules
+verify-modules:
+	@echo "verify-modules target is a placeholder to use same targets as upstream CI"
```
how this calls the target in the other Makefile?

mpryc replied:
It does not, it's just a placeholder that does *nothing*, but the `make ci` target is exactly the same as upstream:

https://github.com/openshift/velero/blob/32e4c959cc1c4e355fc50e69acb6cffc32b74712/Makefile#L330

equals to our Makefile.prow target.

mateusoliveira43 replied:
if it does nothing, while other makefile does anotherthing, it is not the same...

mpryc replied:
Ok will update to call original upstream.

mpryc replied:
Should be resolved. Please check.

mateusoliveira43 replied:
will try to produce a test locally

mpryc commented:
@mateusoliveira43 while adding lint target I did notice that this PR introduced lint error, because we never ran lint jobs, should I fix it with this PR or address with another. We will unfortunately need to backport this to other branches of downstream Velero:

https://github.com/openshift/velero/commit/32e4c959cc1c4e355fc50e69acb6cffc32b74712#diff-23f2d480d6500154fe7ea296a25696240962d708378c7d8f4648958e3579a864R34

The linting error:
```
pkg/repository/udmrepo/kopialib/backend/common_kopia_algorithms_test.go:33:48: unnecessary leading newline (whitespace)
```

mateusoliveira43 commented:
@mpryc first, the errors needs to show in our CI, so adding it to makefile target is needed

We need to confirm the errors are only in our patches, as upstream files should already be linted 

and we need to find a way to always run the same version as upstream. currently it is the same, but after rebase, it will be outdated https://github.com/openshift/release/blob/master/ci-operator/config/openshift/velero/openshift-velero-konveyor-dev.yaml#L5

mpryc commented:
@mateusoliveira43 

> @mpryc 
> and we need to find a way to always run the same version as upstream. currently it is the same, but after rebase, it will be outdated https://github.com/openshift/release/blob/master/ci-operator/config/openshift/velero/openshift-velero-konveyor-dev.yaml#L5

I don't understand that part? We do have go version only in the above file, so after which rebase it will be outdated? We don't run manually container execution of `quay.io/konveyor/builder:ubi9-v1.22.2`, this is hashed and just a comment on the Makefile.prow

mateusoliveira43 commented:
wrong link, sorry

right one https://github.com/vmware-tanzu/velero/blob/main/hack/build-image/Dockerfile#L97

golangci-lint version is my concern, not container

weshayutin commented:
```
ok  	github.com/vmware-tanzu/velero/internal/volume	0.021s	coverage: 48.9% of statements
ok  	github.com/vmware-tanzu/velero/internal/volumehelper	0.026s	coverage: 37.4% of statements
Success!
make[1]: Leaving directory '/home/whayutin/OPENSHIFT/git/OADP/velero'
```
WFM :)

mpryc commented:
> wrong link, sorry
> 
> right one https://github.com/vmware-tanzu/velero/blob/main/hack/build-image/Dockerfile#L97
> 
> golangci-lint version is my concern, not container

How about we do something smart inside a Makefile and if it's not found then fail:
```shell
$ grep 'golangci-lint/master/install.sh' hack/build-image/Dockerfile | grep -oP 'v[0-9]+\.[0-9]+\.[0-9]+'
v1.57.2
```

mateusoliveira43 commented on Makefile.prow:
```diff
@@ -1,92 +1,90 @@
-KUBEBUILDER_ASSETS=$(shell echo $(shell $(GOBIN)/setup-envtest use -p path) | sed 's/ /\\ /g')
-.PHONY: envtest
-envtest: $(GOBIN)/setup-envtest
+.PHONY: lint
```
I would remove this

we are adding without using it

mpryc replied:
I want to introduce prow job for that target, but first we need to have it in the Makefile so prow CI job can be successful. It's chicken-egg problem.

mateusoliveira43 replied:
on OADP, we have everything under unit-test job

would do the same approach here

mpryc replied:
This is very much different from the upstream and usually lint is not part of unit. Having it separate gives clean information that the code is OK, just needs reformatting without checking job logs.

kaovilai commented:
needs rebase

mpryc commented:
> needs rebase

done

sseago commented:
@mpryc I'm concerned that this is going to lead to rebase merge conflict headaches every time upstream Makefile changes. Not sure what the right answer is, though.

mpryc commented:
Corresponding release change: https://github.com/openshift/release/pull/55969

sseago commented:
@kaovilai @mpryc oops. Sorry, I misread the filename. You're right. No concerns here.

weshayutin commented:
/retest

kaovilai commented:
I don't like how this pr is showing 41 commits.. did you fetch openshift before rebasing against openshift/konveyor-dev?

mpryc committed:
Downstream only: Rework of Makefile and incusion of lint

The rework of Makefile to make it more readable and
inclusion of lint as a target as well extract
golangci-lint version from the upstream Dockerfile,
so we test in PROW or locally on the same version as upstream.

Signed-off-by: Michal Pryc <mpryc@redhat.com>

mpryc commented:
> I don't like how this pr is showing 41 commits.. did you fetch openshift before rebasing against openshift/konveyor-dev?

Fixed, that was based before rebase of velero, that's why.

mpryc commented:
/cherry-pick oadp-1.4
